### DISCUSSION TITLE
fatal error: 'stdio.h' file not found\n" on MacOs via lib, but not from CLI

### DISCUSSION THREAD

**Discussion by tgross35** [rust-lang/rust-bindgen#2401] [Q&A] [answered]
Hey everyone,

I'm struggling a bit getting bindgen to pick up on `stdio.h` when used with the library. Similar things have been documented in a few places, like https://github.com/rust-lang/rust-bindgen/issues/1521 - but those seem to always be for C++ rather than C.

The files:

```c
// wrapper.h
#include "stdio.h" // I have also tried <stdio.h>
```

```rs
// build.rs

fn main() {
    const HEADER_PATH: &str = "wrapper.h";
    println!("cargo:rerun-if-changed={HEADER_PATH}");

    let bindings = bindgen::Builder::default()
        .header(HEADER_PATH)
        .clang_args(["-x", "c", "-std=c11"]) // I have tried with and without that, 
        .generate()
        .expect("Unable to generate bindings");

    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("couldn't write bindings");
}
```

Running either of the below works fine:

```sh
$ clang wrapper.h
$ bindgen wrapper.h
```

But running `cargo c` comes up with the following:

```
  --- stderr
  wrapper.h:48:10: fatal error: 'stdio.h' file not found
  thread 'main' panicked at 'Unable to generate bindings: ClangDiagnostic("wrapper.h:48:10: fatal error: 'stdio.h' file not found\n")', 
```

Both `bindgen` and `bindgen-cli` are on 0.63

Does anyone have any suggestions?

**Comment by tgross35** [comment, marked as answer]
After some chasing down, it seems like it's due to my target. Running clang with `-target thumbv7em-none-eabihf` gets the same results

**Reply by mouse07410** [reply]
I'm having the same problem - but in my case Cargo is failing to build a dependency:
- I'm trying to build `rustls`;
- it now uses `aws-lc-rs` crate, which in turn depends on `aws-lc-sys`;
- `aws-lc-sys` fails to build, with the above error from `bindgen` (can't find `stdio.h`).

There's no cross-compilation involved, building on MacOS Sonoma for MacOS Sonoma, on Apple Silicon M2. There's no way (that I know of) to pass `-target` to `clang`.

Please help.

**Reply by mouse07410** [reply]
Experimenting with `build.rs` and passing `--target=aarch64` with `.clang_args()` did not help - same error.

**Comment by mouse07410** [comment]
Apple Silicon M2 Max, MacOS Sonoma 14.3.1, Xcode-15.2, Rust-1.76.0.

`bindgen` consistently fails. Here's the reproducer from your user guide:

```text
$ cargo new bindgen-tst2
     Created binary (application) `bindgen-tst2` package
$ cd bindgen-tst2
$ cargo add --build bindgen
    Updating crates.io index
      Adding bindgen v0.69.4 to build-dependencies.
             Features:
             + logging
             + prettyplease
             + runtime
             + which-rustfmt
             - __cli
             - __testing_only_extra_assertions
             - __testing_only_libclang_16
             - __testing_only_libclang_9
             - experimental
             - static
    Updating crates.io index
$ cargo add env
    Updating crates.io index
      Adding env v0.0.0 to dependencies.
    Updating crates.io index
$ vi Cargo.toml
$ vi wrapper.h
$ vi build.rs
$ cargo c
   Compiling libc v0.2.153
   Compiling proc-macro2 v1.0.78
   Compiling unicode-ident v1.0.12
   Compiling glob v0.3.1
   Compiling prettyplease v0.2.16
   Compiling rustix v0.38.31
   Compiling bitflags v2.4.2
   Compiling memchr v2.7.1
   Compiling cfg-if v1.0.0
   Compiling either v1.10.0
   Compiling minimal-lexical v0.2.1
   Compiling regex-syntax v0.8.2
   Compiling home v0.5.9
   Compiling libloading v0.8.1
   Compiling bindgen v0.69.4
   Compiling itertools v0.12.1
   Compiling lazycell v1.3.0
   Compiling log v0.4.20
   Compiling lazy_static v1.4.0
   Compiling rustc-hash v1.1.0
   Compiling nom v7.1.3
   Compiling shlex v1.3.0
    Checking env v0.0.0
   Compiling clang-sys v1.7.0
   Compiling regex-automata v0.4.5
   Compiling cexpr v0.6.0
   Compiling regex v1.10.3
   Compiling quote v1.0.35
   Compiling syn v2.0.48
   Compiling errno v0.3.8
   Compiling which v4.4.2
   Compiling bindgen-tst2 v0.1.0 (/Users/ur20980/src/bindgen-tst2)
error: failed to run custom build command for `bindgen-tst2 v0.1.0 (/Users/ur20980/src/bindgen-tst2)`
note: To improve backtraces for build dependencies, set the CARGO_PROFILE_DEV_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug information generation.

Caused by:
  process didn't exit successfully: `/Users/ur20980/src/bindgen-tst2/target/debug/build/bindgen-tst2-f9655009e9fd33b8/build-script-build` (exit status: 101)
  --- stdout
  cargo:rustc-link-search=/opt/local/lib
  cargo:rustc-link-lib=bz2
  cargo:rerun-if-env-changed=TARGET
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS_aarch64-apple-darwin
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS_aarch64_apple_darwin
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS
  cargo:rerun-if-changed=wrapper.h

  --- stderr
  wrapper.h:1:10: fatal error: 'bzlib.h' file not found
  thread 'main' panicked at build.rs:25:10:
  Unable to generate bindings: ClangDiagnostic("wrapper.h:1:10: fatal error: 'bzlib.h' file not found\n")
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:645:5
     1: core::panicking::panic_fmt
               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:72:14
     2: core::result::unwrap_failed
               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/result.rs:1649:5
     3: core::result::Result<T,E>::expect
     4: build_script_build::main
     5: core::ops::function::FnOnce::call_once
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
$ 
```

```rs
// build.rs
use std::env;
use std::path::PathBuf;

fn main() {
    // Tell cargo to look for shared libraries in the specified directory
    println!("cargo:rustc-link-search=/opt/local/lib");

    // Tell cargo to tell rustc to link the system bzip2
    // shared library.
    println!("cargo:rustc-link-lib=bz2");

    // The bindgen::Builder is the main entry point
    // to bindgen, and lets you build up options for
    // the resulting bindings.
    let bindings = bindgen::Builder::default()
        // The input header we would like to generate
        // bindings for.
        .header("wrapper.h")
        // Tell cargo to invalidate the built crate whenever any of the
        // included header files changed.
        .parse_callbacks(Box::new(bindgen::CargoCallbacks::new()))
        // Finish the builder and generate the bindings.
        .generate()
        // Unwrap the Result and panic on failure.
        .expect("Unable to generate bindings");

    // Write the bindings to the $OUT_DIR/bindings.rs file.
    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}
```

```c
// wrapper.h
#include <bzlib.h>
```

```toml
# Cargo.toml

[package]
name = "bindgen-tst2"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
env = "0.0.0"

[build-dependencies]
bindgen = "0.69.4"
```

**Comment by mouse07410** [comment]
### Update
Moving `bzlib` out of the picture and adding `stdio.h` to the `wrapper.h` reveals that `bindgen` not only fails to locate header files outside the standard system path, but can't find `stdio.h` as well:

```c
// wrapper.h

#include <stdio.h>
#include <bzlib.h>
```

```rs
// build.rs

use std::env;
use std::path::PathBuf;

fn main() {
    
    // Tell cargo to look for shared libraries in the specified directory
    println!("cargo:rustc-link-search=/opt/local/lib");

    // Tell cargo to tell rustc to link the system bzip2
    // shared library.
    println!("cargo:rustc-link-lib=bz2");

    // The bindgen::Builder is the main entry point
    // to bindgen, and lets you build up options for
    // the resulting bindings.
    let bindings = bindgen::Builder::default()
        // The input header we would like to generate
        // bindings for.
        .header("wrapper.h")
        // Tell cargo to invalidate the built crate whenever any of the
        // included header files changed.
        .parse_callbacks(Box::new(bindgen::CargoCallbacks::new()))
        // Finish the builder and generate the bindings.
        .generate()
        // Unwrap the Result and panic on failure.
        .expect("Unable to generate bindings");

    // Write the bindings to the $OUT_DIR/bindings.rs file.
    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}
```

Reproduces the failure:
```sh
$ bindgen wrapper.h -- -v
clang version 16.0.6
Target: arm64-apple-darwin23.3.0
Thread model: posix
InstalledDir: 
ignoring nonexistent directory "lib/clang/16/include"
ignoring nonexistent directory "/usr/include"
ignoring duplicate directory "/usr/local/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/include
 /opt/local/libexec/llvm-16/lib/clang/16/include
 /System/Library/Frameworks
 /Library/Frameworks
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
End of search list.
wrapper.h:3:10: fatal error: 'stdio.h' file not found
panicked at /Users/ur20980/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bindgen-cli-0.69.4/main.rs:52:36:
Unable to generate bindings: ClangDiagnostic("wrapper.h:3:10: fatal error: 'stdio.h' file not found\n")
$ 
$ port select clang
Available versions for clang:
	apple-clang
	mp-clang-16
	mp-clang-17 (active)
	none
$ 
```

Observations:
- `bindgen` uses the wrong Clang;
- `bindgen` ignores the SDK locations on MacOS (both Xcode-provided, and CommandLineTools-provided);
- `bindgen` does not have an option to tell it (especially via env vars) where to look for header files and SDK.

### Update 2
I found two workarounds:

1. **If** you have control of the relevant (`bindgen`-invoking) `build.rs`, then adding to the `let bindings = bindgen::Builder::default()` the following stanza `.clang_args(["-I/opt/local/include", "-I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"])`;
2. Getting rid of Clang-16 completely, deleting/erasing it from your system, rebuilding everything that depended on it with Apple Clang or Macports Clang-17, **and** rebuilding `bindgen`.

I still think `bindgen` should be fixed to address the path to SDK headers on MacOS correctly.

**Reply by pvdrz** [reply]
> `bindgen` uses the wrong Clang

This can be configured via: https://github.com/KyleMayes/clang-sys?tab=readme-ov-file#environment-variables. clang-sys also uses the higher version of clang by default. Meaning that they are not being able to find the "right" clang.

> `bindgen` ignores the SDK locations on MacOS (both Xcode-provided, and CommandLineTools-provided);

This is more of a clang issue in my opinion. Bindgen doesn't try to guess where your system headers are located and just calls clang. It would be more reasonable if clang included such paths instead.

> `bindgen` does not have an option to tell it (especially via env vars) where to look for header files and SDK.

Basically, the same answer as before. Bindgen doesn't look for header files. The only thing bindgen can do is pass extra arguments to clang as you suggested. There are also the `BINDGEN_EXTRA_CLANG_ARGS` and `BINDGEN_EXTRA_CLANG_ARGS_<TARGET>` environment variables, so you can do this without having to modify the build-script
